File: day06/src/main.go
package main

import (
	_ "embed"
	"fmt"
	"strconv"
	"strings"
)

//go:embed input.txt
var input string

func simulate(fishes [9]int, days int) int {
	for i := 0; i < days; i++ {
		pregnant := fishes[0]
		for j := 1; j < 9; j++ {
			fishes[j-1] = fishes[j]
		}
		fishes[8] = pregnant
		fishes[6] += pregnant
	}
	sum := 0
	for _, count := range fishes {
		sum += count
	}
	return sum
}

func main() {
	lines := strings.Split(input, ",")
	fishes := [9]int{0}
	for _, line := range lines {
		i, _ := strconv.Atoi(line)
		fishes[i]++
	}
	fmt.Println("Part 1:", simulate(fishes, 80))
	fmt.Println("Part 2:", simulate(fishes, 256))
}

File: day06/src/main.rs
const GESTATION: usize = 7;
const FIRST_CYCLE_DELAY: usize = 2;
const POP_SIZE: usize = GESTATION + FIRST_CYCLE_DELAY;

fn simulate(mut fishes: [usize; POP_SIZE], days: usize) -> usize {
    for _ in 0..days {
        // All the pregnant fish i.e. about to produe a new lanternfish
        let pregnant = fishes[0];
        // Each fish de-ages one day
        for i in 1..POP_SIZE {
            fishes[i - 1] = fishes[i];
        }
        // Each pregnant fish has exactly one baby which starts the cycle
        // at 8 days old
        fishes[POP_SIZE - 1] = pregnant;
        // Each pregnant fish then goes back to 6 days old
        fishes[GESTATION - 1] += pregnant;
    }
    // How many fish total
    fishes.iter().sum()
}

fn main() {
    let mut fishes = [0usize; POP_SIZE];
    for age in include_str!("input.txt").split(",") {
        fishes[age.parse::<usize>().unwrap()] += 1;
    }
    println!("Part 1: {}", simulate(fishes.clone(), 80));
    println!("Part 2: {}", simulate(fishes, 256));
}
